source/extensions/filters/http/lua/lua_filter.cc

```diff
@@ -629,21 +632,38 @@ int StreamHandleWrapper::luaTimestamp(lua_State* state) {
   auto now = time_source_.systemTime().time_since_epoch();
 
   absl::string_view unit_parameter = luaL_optstring(state, 2, "");
+  auto resolution = getTimestampResolution(unit_parameter);
+  if (resolution == Timestamp::Resolution::Millisecond) {
+    auto milliseconds_since_epoch =
+        std::chrono::duration_cast<std::chrono::milliseconds>(now).count();
+    std::string timestamp = std::to_string(milliseconds_since_epoch);
+    lua_pushstring(state, timestamp.c_str());
+  } else if (resolution == Timestamp::Resolution::Microsecond) {
+    auto microseconds_since_epoch =
+        std::chrono::duration_cast<std::chrono::microseconds>(now).count();
+    std::string timestamp = std::to_string(microseconds_since_epoch);
+    lua_pushstring(state, timestamp.c_str());
+  } else {
+    luaL_error(state, "timestamp format must be MILLISECOND or MICROSECOND.");
+  }
+  return 1;
```

Comment:
Why we update the timestamp method to return a string? Is there some risks of overflow?

Reply:
According to [cppreference](https://en.cppreference.com/w/cpp/chrono/duration,) for microsecond resolution we need "signed integer type of at least 55 bits*.
If we push this with lua_pushnumber(), it gets converted to a double which according to [wikipedia](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) has a precision of 53 bits.
So I see the possibility of something going wrong, even though I'm not an expert on these types of conversion.

A duration, regardless of resolution, is an integer since epoch time. According to my limited lua knowledge, an integer of that kind can't be pushed to lua. So it seems simpler to provide the integer as a string and have the caller decide what to do with it.

Any advice on how to tackle this in a smarter way is mostly welcomed.

Reply:
Thanks. Precision sounds a reasonable reason. But it seems that this still a break change of lua API. It will changes return type of timestamp call.

May be a runtime guard is necessary here? cc @mattklein123

Reply:
I don't think we can change this. I would add a new API that provides the microsecond resolution and leave the existing API the same. Thank you!

Reply:
Thank you for your input! 
I've implemented a new function timestampString(), so that we can have either milli/microsecond timestamp in string representation. In that way, we can bypass conversions to double etc.It might be useful for those cases that we only want to print/log the timestamp instead of doing arithmetic operations with it. What's your opinion on this?
Otherwise I can make a new microsecondTimestamp() to return only the microsecond resolution timestamp.